packages/core/src/client.ts

```diff
@@ -692,17 +693,36 @@ export class MedplumClient extends EventTarget {
 
     if (options?.accessToken) {
       this.setAccessToken(options.accessToken);
+      this.initPromise = Promise.resolve();
+    } else if ((this.storage as AsyncBackedClientStorage).initialized !== undefined) {
```

Comment:
Structurally, this is clever, and I think it nicely solves a really tricky problem.

This was already a pretty gnarly hot spot in MedplumClient, full of yucky complexity, so if it's ok with you I'd like us to be extra thoughtful about naming and API design.

Given that `storage` is already a custom Medplum type, how would you feel about making `getInitPromise()` an optional method on the existing interface?  Then we wouldn't need the extra type, the extra `as`, etc.

Reply:
I think optional methods are kind of weird for interfaces, having type hints for whether or not a method actually exists on a class is pretty good when possible IMO.

This condition is the one case where there is ambiguity since we aren't sure if the user passed in an `AsyncBackedClientStorage` or just a normal `ClientStorage`, but in all other circumstances it would be good to have proper typing on a class creating implementing `AsyncBackedClientStorage` vs just normal `ClientStorage` which should not have the method.